Remove database records when image is deleted from media gallery

coderimus opened the pull request:
### Description
#582 Remove database records when an image is deleted from media gallery.

This PR provides delete functionality for the media asset and related data. Also, the asset object got the CASCADE on delete instruction when the media object was triggered to be deleted.

Provides some refactoring and code improvements like logging, logical improvements in the code execution etc

### Fixed Issues
N/A

### Manual testing scenarios
1. Login to admin panel
1. Open Magento Media Gallery (i.e. go to Cms -> Pages, edit the page and insert image)
1. Click "Search Adobe Stock" button to open images grid
1. Click on any image to expand preview section
1. Save Preview version of the image
1. Delete the saved image in the media gallery

### Acceptance criteria
`media_gallery_asset` and a`dobe_stock_asset` tables do not contain deleted image information

coderimus committed:
Remove media asset data in storage on image delete

coderimus commented:
Hello @sivaschenko ,
Please create an issue for coverage with unit tests the DeleteById class and assign on me. Thank you :)

coderimus commented:
@sivaschenko please, ignore my previous message about the new issue for cover the delete by id functionality. I added this test in scope of this PR.

2 moments to highlight:

1. The DeleteById unit test covers 2 scenarios: when the delete function executed successfully and when some error occurred.
 1. You will see this line `$connection->delete($tableName, 'id = ' . $mediaAsset->getId());` I choose concatenation instead of sending `where` query parameters as an array because after the deep investigation I found that this approach with an array makes us completely depended on the Zend implementation - it accept mixed types for where condition and convert array to string at the low level. I would prefer to send a string to the delete method as a where parameter.

sivaschenko commented on AdobeMediaGallery/Model/Asset/Command/DeleteById.php:
```diff
@@ -0,0 +1,83 @@
+            /** @var AdapterInterface $connection */
+            $connection = $this->resourceConnection->getConnection();
+            $tableName = $this->resourceConnection->getTableName(self::TABLE_MEDIA_GALLERY_ASSET);
+            $connection->delete($tableName, 'id = ' . $mediaAsset->getId());
```
I saw you comment, but I think it's still necessary to pass the argument as an array as it is a good practice to quote all variables used for Mysql queries. To be sure to be sure :)

sivaschenko replied:
Also, ideally `id` should be a constant as well

sivaschenko commented on AdobeMediaGallery/Model/Asset/Command/DeleteById.php:
```diff
@@ -0,0 +1,83 @@
+            $tableName = $this->resourceConnection->getTableName(self::TABLE_MEDIA_GALLERY_ASSET);
+            $connection->delete($tableName, 'id = ' . $mediaAsset->getId());
+        } catch (\Exception $exception) {
+            $message = __(
```
Can you please check why static tests are failing for this construction:

    inconsistency the number of arguments and placeholders were discovered: 

    Phrase: 'Could not delete media asset with id %id: %error' 
    File: /var/www/html/app/code/Magento/AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php 
    Line: 151

sivaschenko commented on AdobeMediaGallery/Model/Asset/Command/GetListByPath.php:
```diff
@@ -0,0 +1,85 @@
+            $connection = $this->resourceConnection->getConnection();
+            $select = $connection->select()
+                ->from(self::TABLE_MEDIA_GALLERY_ASSET)
+                ->where('path in (?)', $mediaFilePath);
```
I'd move `path` to a constant

sivaschenko commented on AdobeMediaGallery/Model/Asset/Command/GetListByPath.php:
```diff
@@ -0,0 +1,85 @@
+
+            $mediaAssets = [];
+            foreach ($data as $id => $assetData) {
+                $mediaAssets[$id] = $this->mediaAssetFactory->create(['data' => $assetData]);
```
Cool approach ;)

sivaschenko commented on AdobeMediaGallery/Model/Asset/Command/GetListByPath.php:
```diff
@@ -0,0 +1,85 @@
+     * @return array
+     * @throws IntegrationException
+     */
+    public function execute(string $mediaFilePath): array
```
Considering that path is unique, I'd rename the command to `GetByPath` and return `AssetInterface`

sivaschenko commented on AdobeMediaGallery/Plugin/Product/Gallery/Processor.php:
```diff
@@ -0,0 +1,85 @@
+     */
+    public function afterRemoveImage(ProcessorSubject $subject, $result, Product $product, $file)
+    {
+        if (!is_string($file)) {
```
I like this attention to details 👍

sivaschenko commented on AdobeMediaGallery/Plugin/Product/Gallery/Processor.php:
```diff
@@ -0,0 +1,85 @@
+
+        /** @var AssetInterface $mediaAsse */
+        foreach ($mediaAssetList as $mediaAsset) {
+            $this->deleteMediaAssetById->execute($mediaAsset->getId());
```
I was thinking that `DeleteByPath` command can replace both `DeleteById` and `GetListByPath`

sivaschenko commented on AdobeMediaGallery/Plugin/Wysiwyg/Images/Storage.php:
```diff
@@ -0,0 +1,104 @@
+
+        $relativePath = $this->filesystem->getDirectoryRead(DirectoryList::MEDIA)->getRelativePath($target);
+        $mediaAssetList = $this->getMediaListByPath->execute($relativePath);
+        if (!isset($mediaAssetList)) {
```
It's impossible for these conditions to result in `true` considering the current implementation

sivaschenko commented on AdobeMediaGallery/Plugin/Wysiwyg/Images/Storage.php:
```diff
@@ -0,0 +1,104 @@
+            foreach ($mediaAssetList as $mediaAsset) {
+                $this->deleteMediaAssetById->execute($mediaAsset->getId());
+            }
+        } catch (\Exception $exception) {
```
It will make sense to try-catch wrap the delete operation in the Processor plugin as well

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+        );
+
+        $this->mediaAsset = $this->createMock(AssetInterface::class);
+        $this->adapter = $this->createMock(AdapterInterface::class);
```
If the `adaptor` is created in `setUp` - it's a good idea to configure the `resourceConnection` mock to return this adaptor in `setUp` method as well.

In this case we don't need a property for `resourceConnection` and will work only with `adaptor` property further

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+     */
+    public function testSuccessfulDeleteByIdExecution(int $mediaAssetId)
+    {
+        $this->getMediaAssetById->expects($this->at(0))
```
```suggestion
        $this->getMediaAssetById->expects($this->once())
```

coderimus committed:
Merge branch 'develop' into 582-remove-database-records-on-image-remove

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+            ->with($mediaAssetId)
+            ->willReturn($this->mediaAsset);
+
+        $this->resourceConnection->expects($this->at(0))
```
```suggestion
        $this->resourceConnection->expects($this->once())
```

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+            ->willReturn($this->adapter);
+
+        $tableName = 'media_gallery_asset';
+        $this->resourceConnection->expects($this->at(1))
```
I don't think it's necessary to test that `getTableName` is called strictly after `getConnection`. While that sequence of calls is not required, asserting the sequence makes the test more fragile
```suggestion
        $this->resourceConnection->expects($this->once())
```

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+        $this->resourceConnection->expects($this->at(1))
+            ->method('getTableName')
+            ->with($tableName)
+            ->willReturn($tableName);
```
I'd use different table names to ensure that the correct one is used further

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+            ->method('getId')
+            ->willReturn($mediaAssetId);
+
+        $this->adapter->expects($this->at(0))
```
There is only one call to `adapter::delete` method expected, no need to verify the sequence
```suggestion
        $this->adapter->expects($this->once())
```

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+            ]
+        );
+
+        $this->mediaAsset = $this->createMock(AssetInterface::class);
```
I'd create the asset mock directly in the test as this is a dynamic entity (theoretically more than one asset can be used)

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+     */
+    public function testExceptionOnDeleteExecution(int $mediaAssetId)
+    {
+        $this->getMediaAssetById->expects($this->at(0))
```
Please use `at` method only when you need to verify the sequence of calls (or when the same method is called several times with either different arguments or different return values)

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+            ->method('critical')
+            ->with($testMessage)
+            ->willReturnSelf();
+        $this->expectExceptionMessage($testMessage);
```
I wouldn't verify the test message. It will make the test fail even if small correction to exception message phrasing is applied

sivaschenko commented on AdobeMediaGallery/Test/Unit/Model/Asset/Command/DeleteByIdTest.php:
```diff
@@ -0,0 +1,177 @@
+    /**
+     * @return array
+     */
+    public function deleteMediaAssetTestScopes(): array
```
Data provider is not necessary to providing id's in my opinion, the behaviour is pretty much the same for any `id`.

Also, (note for the future unit tests) data providers are usually used to provide at least two arguments for the test:
test data and expected result (or flags telling the test how the verification should change depending on test data provided) i.e. `$input, $output` or `$input, $isExceptionExpected`

sivaschenko commented on AdobeStockAsset/etc/db_schema.xml:
```diff
@@ -27,7 +27,7 @@
         <constraint xsi:type="foreign" referenceId="ADOBE_STOCK_ASSET_CATEGORY_ID_ADOBE_STOCK_CATEGORY_ID" table="adobe_stock_asset" column="category_id" referenceTable="adobe_stock_category" referenceColumn="id" onDelete="SET NULL"/>
         <constraint xsi:type="foreign" referenceId="ADOBE_STOCK_ASSET_CREATOR_ID_ADOBE_STOCK_CREATOR_ID" table="adobe_stock_asset" column="creator_id" referenceTable="adobe_stock_creator" referenceColumn="id" onDelete="SET NULL"/>
-        <constraint xsi:type="foreign" referenceId="ADOBE_STOCK_ASSET_MEDIA_GALLERY_ID_ADOBE_STOCK_CREATOR_ID" table="adobe_stock_asset" column="media_gallery_id" referenceTable="media_gallery_asset" referenceColumn="id" onDelete="SET NULL"/>
+        <constraint xsi:type="foreign" referenceId="ADOBE_STOCK_ASSET_MEDIA_GALLERY_ID_ADOBE_STOCK_CREATOR_ID" table="adobe_stock_asset" column="media_gallery_id" referenceTable="media_gallery_asset" referenceColumn="id" onDelete="CASCADE"/>
```
I think `CREATOR_ID` was added to the constraint name by mistake before, it would be good to correct the name

sivaschenko commented on AdobeMediaGalleryApi/Model/Asset/Command/SaveInterface.php:
```diff
@@ -11,19 +11,18 @@
- * Save media asset.
- *
+ * Save media asset
  * @api
```
Could you please remove all the `@api` annotation from `AdobeMediaGalleryApi` module. We cannot add API to the patch release, but we can mark these interfaces back with `@api` later in 2.4 release.

I missed this point initially, but thankfully that was brought up by @joni-jones during the public meeting yesterday

sivaschenko reviewed (changes requested):
@coderimus Thanks for the pull request! Looks great!

As per your request, I did a very detailed and critical review of the unit tests :)

Please see my review comments

sivaschenko commented on AdobeMediaGallery/Plugin/Wysiwyg/Images/Storage.php:
```diff
@@ -0,0 +1,104 @@
+    /**
+     * @var Filesystem
+     */
+    protected $filesystem;
```
Magento Health Index test results helped to identify this unnecessary protected scope
```suggestion
    private $filesystem;
```

coderimus committed:
Chagnes according review

coderimus committed:
Merge latest develop

coderimus committed:
Fix Magento static tests

coderimus commented:
@sivaschenko thank you for your detail and constructive review. I implemented the requested changes, but there are two questions:

1. You wrote about [those two lines of the test class](https://github.com/magento/adobe-stock-integration/pull/597/files#diff-b0139084384152b31004be7e37c8dae7R83-R84): "I'd use different table names to ensure that the correct one is used further" Sorry, but I didn't catch the idea of this suggestion. If I used different table names, I would get an error. Please, describe a little bit detail what you mean :)

1. I fixed some static tests for error message parameters inconsistent, but there are some more: AdobeMediaGallery has soft dependencies of Magento_CMS and Magento_Catalog. I added the sequence to the module, but there are still errors in the static tests. Can you please check the static test report? Maybe we can ignore those messages or turnoff them for media module?

sivaschenko committed:
Merge branch 'develop' into 582-remove-database-records-on-image-remove

sivaschenko commented on AdobeMediaGallery/Model/Asset/Command/GetByPath.php:
```diff
@@ -0,0 +1,90 @@
+            $connection = $this->resourceConnection->getConnection();
+            $select = $connection->select()
+                ->from(self::TABLE_MEDIA_GALLERY_ASSET)
+                ->where(self::MEDIA_GALLERY_ASSET_PATH . ' in (?)', $mediaFilePath);
```
```suggestion
                ->where(self::MEDIA_GALLERY_ASSET_PATH . ' = ?', $mediaFilePath);
```

sivaschenko commented on AdobeMediaGallery/Model/Keyword/Command/SaveAssetKeywords.php:
```diff
@@ -47,7 +47,11 @@ public function __construct(
+     *
+     * @param KeywordInterface[] $keywords
+     * @param int $assetId
+     * @throws CouldNotSaveException
```
I'd use `@inheritdoc` here

coderimus replied:
I removed it because of [this rule in the dockbloc standards](https://devdocs.magento.com/guides/v2.3/coding-standards/docblock-standard-general.html#inheritdoc) "The @inheritdoc tag SHOULD NOT be used. If a child class method requires a long description to explain its purpose, it may use @inheritdoc to indicate the new description is intended as an addition to the parent method description. In general such method overrides are a code smell and should be used as an incentive to make the code more self-documenting if possible."

sivaschenko commented on AdobeMediaGallery/etc/module.xml:
```diff
@@ -6,5 +6,10 @@
-    <module name="Magento_AdobeMediaGallery" />
+    <module name="Magento_AdobeMediaGallery">
+        <sequence>
+            <module name="Magento_Catalog"/>
```
moule.xml is used to define installation dependencies. `AdobeMediaGallery` depends on Catalog and CMS modules, but they do not necessarily need to be installed before `AdobeMediaGallery` (as there are not database schema dependency). So these dependencies should only be added to composer.json

coderimus replied:
@sivaschenko thank you for an explanation! I got the point now with this soft/hard dependencies in our case :)

sivaschenko committed:
magento/adobe-stock-integration#597: Fixed code review comments

sivaschenko committed:
magento/adobe-stock-integration#597: Fixed static tests

ghost commented:
Hi @coderimus, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento+adobe-stock-integration+597), it will take less than a minute. </br>Your feedback will help us to improve contribution process.
